### PR TITLE
feat: Check for auth failures in liveness and readiness probes

### LINKED ISSUE #3351: Chnages in IRSA Role does not restart Karpenter Pods which causes AWS Auth failures
### Version

__Karpenter Version:__ v0.22.1

__Kubernetes Version:__ v1.24

### Expected Behavior

Kaprneter should check if the Service account is modified with EKS IAM Role then it should restart pods to get new STS Credentials. Which causes AWS auth to fail

### Actual Behavior

IT uses the Old STS Creds which are expired because the SA a=was modified to use new role 

### Steps to Reproduce the Problem

Just Chnage the karpeneter SA to use new IRSA Role and u will start to see AWS Auth failures in logs

### Resource Specs and Logs

ERROR	controller.provisioner	creating scheduler, getting instance types, fetching instance types using ec2.DescribeInstanceTypes, AuthFailure: AWS was not able to validate the provided access credentials

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by kevinmingtarja**
Fixes #3351 

**Description**

**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ellistarn** [comment]
Can you include a description for this PR?

File: pkg/utils/errcode/errcode.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package errcode
+
+const (
+    DryRunOperation string = "DryRunOperation"
```
Comment (jonathan-innis) [inline pkg/utils/errcode/errcode.go:18]: There's an `errors` package. We could consider moving this error code into that package with the others

File: pkg/providers/subnet/subnet.go [hunk 2]
```diff
@@ -201,11 +204,22 @@ func (p *Provider) UpdateInflightIPs(createFleetInput *ec2.CreateFleetInput, cre
...
+		return err
+	}
 	p.Unlock()
 	return nil
 }
 
+func (p *Provider) checkAWSAuth(ctx context.Context) error {
+	_, err := p.ec2api.DescribeSubnetsWithContext(ctx, &ec2.DescribeSubnetsInput{DryRun: aws.Bool(true)})
+	var aerr awserr.Error
+	if errors.As(err, &aerr) && aerr.Code() == errcode.DryRunOperation {
```
Comment (jonathan-innis) [inline pkg/providers/subnet/subnet.go:217]: Maybe consider creating a `IsDryRunOperationError` and `IgnoreDryRunOperationError` similar to the funcs provided in the client-go errors package.

File: pkg/providers/subnet/subnet.go [hunk 3]
```diff
@@ -201,11 +204,22 @@ func (p *Provider) UpdateInflightIPs(createFleetInput *ec2.CreateFleetInput, cre
 
 func (p *Provider) LivenessProbe(req *http.Request) error {
 	p.Lock()
-	//nolint: staticcheck
+	if err := p.checkAWSAuth(req.Context()); err != nil {
```
Comment (jonathan-innis) [inline pkg/providers/subnet/subnet.go:207]: Most likely, I don't think this needs to be locked by a mutex.

File: pkg/providers/pricing/pricing.go [hunk 4]
```diff
@@ -387,11 +389,22 @@ func (p *Provider) UpdateSpotPricing(ctx context.Context) error {
 func (p *Provider) LivenessProbe(req *http.Request) error {
 	// ensure we don't deadlock and nolint for the empty critical section
 	p.mu.Lock()
-	//nolint: staticcheck
+	if err := p.checkAWSAuth(req.Context()); err != nil {
```
Comment (jonathan-innis) [inline pkg/providers/pricing/pricing.go:392]: Most likely, I don't think this needs to be locked by a mutex.

File: pkg/providers/pricing/pricing.go [hunk 5]
```diff
@@ -387,11 +389,22 @@ func (p *Provider) UpdateSpotPricing(ctx context.Context) error {
...
+		return err
+	}
 	p.mu.Unlock()
 	return nil
 }
 
+func (p *Provider) checkAWSAuth(ctx context.Context) error {
+	_, err := p.ec2.DescribeSpotPriceHistoryWithContext(ctx, &ec2.DescribeSpotPriceHistoryInput{DryRun: aws.Bool(true)})
+	var aerr awserr.Error
+	if errors.As(err, &aerr) && aerr.Code() == errcode.DryRunOperation {
```
Comment (jonathan-innis) [inline pkg/providers/pricing/pricing.go:402]: Maybe consider creating a `IsDryRunOperationError` and `IgnoreDryRunOperationError` similar to the funcs provided in the client-go errors package.

File: pkg/providers/instancetype/instancetype.go [hunk 6]
```diff
@@ -127,6 +133,15 @@ func (p *Provider) LivenessProbe(req *http.Request) error {
 	return nil
 }
 
+func (p *Provider) checkAWSAuth(ctx context.Context) error {
+	_, err := p.ec2api.DescribeInstanceTypesWithContext(ctx, &ec2.DescribeInstanceTypesInput{DryRun: aws.Bool(true)})
+	var aerr awserr.Error
+	if errors.As(err, &aerr) && aerr.Code() == errcode.DryRunOperation {
```
Comment (jonathan-innis) [inline pkg/providers/instancetype/instancetype.go:139]: Maybe consider creating a `IsDryRunOperationError` and `IgnoreDryRunOperationError` similar to the funcs provided in the client-go errors package.

File: pkg/operator/operator.go [hunk 7]
```diff
@@ -175,7 +176,7 @@ func withUserAgent(sess *session.Session) *session.Session {
 func checkEC2Connectivity(ctx context.Context, api *ec2.EC2) error {
```
Comment (jonathan-innis) [inline pkg/operator/operator.go:176]: Do we need this function check in the beginning now if the liveness/readiness is doing this now?

**Comment by jonathan-innis** [comment]
Awesome work, thanks for the contribution! 🎉  Just some comments and some questions to settle on

**Comment by kevinmingtarja** [comment]
Thanks for the review @jonathan-innis! I'll take a look at it again some time next week. I've been caught up in work lately 😅

**Pull request closed without merging**